internal/pkg/upload/upload.go

```diff
@@ -0,0 +1,341 @@
+					<-chunkT.C
+				}
+				u.cancel(id)
+				return
+			case <-chunkT.C: // no chunk progress within chunk timer, expire operation
+				log.Trace().Str("uploadID", id).Msg("upload operation chunk activity timed out")
+				// stop and drain total timer
+				if !total.Stop() {
+					<-total.C
+				}
+				u.cancel(id)
+				return
+			case <-chunkRecv: // chunk activity, update chunk timer
+				if !chunkT.Stop() {
+					<-chunkT.C
+				}
+				chunkT.Reset(chunkProgressTimeout)
+			case <-complete: // upload operation complete, clean up
+				if !chunkT.Stop() {
+					<-chunkT.C
+				}
+				if !total.Stop() {
+					<-total.C
+				}
+				u.finalize(id)
+				return
+			}
+		}
+	}()
+	info := Info{
+		ID:        id,
+		DocID:     docID,
+		ChunkSize: MaxChunkSize,
+		Source:    source,
+		Total:     size,
+		Hasher:    hasher,
+		HashSum:   hashsum,
+	}
+	cnt := info.Total / info.ChunkSize
+	if info.Total%info.ChunkSize > 0 {
+		cnt += 1
+	}
+	info.Count = int(cnt)
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	u.current[id] = upload{
+		opToken:       token,
+		chunkThrottle: throttle.NewThrottle(u.parallelChunkLimit),
+		complete:      complete,
+		chunkRecv:     chunkRecv,
+		Info:          info,
+	}
+	return info, nil
+}
+
+func (u *Uploader) Chunk(uplID string, chunkID int) (ChunkInfo, error) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	upl, valid := u.current[uplID]
+	if !valid {
+		return ChunkInfo{}, ErrInvalidUploadID
+	}
+	if chunkID < 0 || chunkID >= upl.Info.Count {
+		return ChunkInfo{}, errors.New("invalid chunk number")
+	}
+
+	token := upl.chunkThrottle.Acquire(strconv.Itoa(chunkID), time.Hour)
+	if token == nil {
+		return ChunkInfo{}, ErrMaxConcurrentUploads
+	}
+	upl.chunkRecv <- struct{}{}
+	if !upl.begun {
+		upl.begun = true
+	}
+	u.current[uplID] = upl
+
+	return ChunkInfo{
+		ID:            chunkID,
+		FirstReceived: upl.begun,
+		Final:         chunkID == upl.Info.Count-1,
+		Upload:        upl.Info,
+		Token:         token,
+	}, nil
+}
+
+func (u *Uploader) Complete(id string, bulker bulk.Bulk) (Info, error) {
+	info, valid := u.current[id]
+	if !valid {
+		return Info{}, ErrInvalidUploadID
+	}
+
+	ok, err := u.allChunksPresent(info.Info, bulker)
+	if err != nil {
+		return Info{}, err
+	}
+	if !ok {
+		return Info{}, ErrMissingChunks
+	}
+
+	ok, err = u.verifyChunkData(info.Info, bulker)
+	if err != nil {
+		return Info{}, err
+	}
+	if !ok {
+		return Info{}, errors.New("file contents did not pass validation")
+	}
+
+	// @todo: verify chunks
+	// verify hashes, etc
+	u.current[id].complete <- struct{}{}
+	return info.Info, nil
+}
+
+func (u *Uploader) cleanupOperation(uplID string) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	if upload, ok := u.current[uplID]; ok {
+		if upload.opToken != nil {
+			upload.opToken.Release()
+		}
+	}
+	delete(u.current, uplID)
+}
+
+func (u *Uploader) cancel(uplID string) error {
+	u.cleanupOperation(uplID)
+
+	// @todo: delete any uploaded chunks from ES
+	// leave header doc and mark failed?
```

Comment:
How important is this? Is it possible that orphaned chunks could accumulate in ES over and over again because uploads from an agent on a poor network connection can never complete?

Reply:
less important now that these indices have an ILM policy that cleans up and deletes documents on a somewhat short timescale (XX days)